[WIP] Adding a througput chart to the benchmark summary.

gz opened the pull request:
A first version to address  #149. This adds a throughput chart to the summary page.

Some more thoughts:

It was a bit of a nuisance to get the Throughput struct down to the reporting and plotting, maybe there is an easier, better way that I don't know?

Also, plot::line_comparison_throughput is doing unwrap on Option<Throughput> so the caller should check that every BenchmarkId has a throughput (and that its of the same type probably). I can add that if the general direction of this pull request is deemed fine.

Maybe, instead of a line-plot a barchart will work better here?

To test I use: 
```
cargo bench -- from_elem
open target/criterion/report/index.html
```

gz commented:
It seems I got confused by the two structs report::BenchmarkId which has throughput and BenchmarkId which doesn't store throughput in benchmark_group.rs which I also mistakenly assumed I was passing around. 
So I can simplify this pull request quite a bit.

bheisler commented:
Hey, sorry I haven't had a chance to respond to this yet. Yeah, `report::BenchmarkID` is the way to go. I should also mention there's also a `ValueFormatter::scale_for_throughput` that I added specifically to support this; I haven't looked at the code changes yet so you may have seen that already.

bheisler commented on src/plot/summary.rs:
```diff
@@ -121,20 +121,116 @@ pub fn line_comparison(
+                let mut y = [Sample::new(sample).mean()];
+                let tput = tput.clone().unwrap();
+
+                error!("{:?}", y[0]);
```
Leftover debugging output.

bheisler reviewed (changes requested):
As you noticed, you can get the throughput out of the benchmark ID so all of the changes made to pass the throughput along are unnecessary. There's also some leftover debug output.

Otherwise this looks good. It won't make it into 0.3.0 (which I plan to release today) but it won't require a breaking change so we can add it to 0.3.1.

gz committed:
Adds ability to plot throughput on summary page.

Fixes #149.

gz commented on src/html/mod.rs:
```diff
@@ -801,6 +802,39 @@ impl Html {
+            .collect();
+        let mut line_tput_path = None;
+
+        if value_types.iter().all(|x| x == &value_types[0])
```
Does this if() cover all cases?

bheisler replied:
I think this test will return false if there are multiple different throughput values. For example, some benchmarks could have a throughput of `Bytes(1000u64)` and others could be `Bytes(500u64)`, and that would register as different throughput types. I think what you want is to check if they're all either `Bytes(_)` or `Elems(_)`. In that case, I think something like this should work:

```
let all_throughputs_match = value_types.iter().all(|x| x == &value_types[0]) && throughput_types[0].is_some();
if all_throughputs_match { ... }
```

You don't need the third clause because if all of the value-types match then the throughputs must all have the same variant. Maybe add a comment explaining that though, it's not obvious from reading the code.

gz commented:
Thanks, I updated this pull request, let's see if it passes CI.

bheisler commented on src/plot/summary.rs:
```diff
@@ -121,6 +121,106 @@ pub fn line_comparison(
+    title: &str,
+    all_curves: &[&(&BenchmarkId, Vec<f64>)],
+    path: &str,
+    value_type: ValueType,
```
Taking the value type as a parameter here is unnecessary; you can get that information from the throughputs in the BenchmarkIDs. Removing it also allows you to simplify the calling code a bit.

bheisler commented on src/plot/summary.rs:
```diff
@@ -121,6 +121,106 @@ pub fn line_comparison(
+                .set(axis_scale.to_gnuplot())
+        });
+
+    let mut i = 0;
```
Move this down to the loop where it's used.

bheisler reviewed (changes requested):
Better! Just some minor things to clean up.

gz committed:
Fix issues from review in #315.

gz commented:
Finally got around to addressing the reviews.

The build currently fails due to unrelated issues,should be fixed with #336

gz committed:
Fix throughput check and make plot a scatter instead of line plot.

gz committed:
Include ParameterizedBenchmark config string in plot label.

gz committed:
Plot label takes function parameterization into account.

gz commented:
I changed the plotting for throughput slightly to take into account parameterized benchmarks.

gz committed:
Put deny warnings back.

gz committed:
Make unit decision only based on typical value.

Otherwise the summary plot may end up with different unit display as
the #elements can change for different benchmark.

gz committed:
Undeny warnings to avoid OsRng error.

gz committed:
Make throughput unit decision only based on typical_value.

Otherwise this leads to problem in graphs where we can have
different values in Throughput::Elements, or Throughput::Bytes
that lead to putting values in the same plot with different units.

However, after changing this I had to adjust the format_throughput
function because it would just use the raw time as a typical value
before which could lead to odd unit choices (like 2_000_000 Kelem/s).
With these changes, now the format_throughput also takes a typical value
that can be estimated in advance with Throughput::per_second().

gz committed:
Fix plot to have a line similar to line chart.

Also finds the maximum tput for the plot properly by looking at
tput values and not just mean.

Ekleog commented:
Hey! Just wanted to ask if there's any news on this PR, as it'd be very nice :)

gz commented:
Hey, the branch/PR works/worked for me -- feel free to give it a try ... Unfortunately, I couldn't quite fit my requirements for a multi-core benchmark harness into criterion (even with `iter_custom`) so I am currently not actively relying on the patch but maybe @bheisler could still check if it's something for master if it's useful for you.

h33p commented:
It would be really nice to have this in. Although it fails to generate a plot for logarithmic scale (min yrange can't be zero). That should can be solved by removing `set(Range::Limits(0, max_formatted[0])` (line 181 from summary.rs). If this is still of interest to @bheisler I could take over this and finish implementing the feature.

bheisler commented:
Hey folks, thanks for your patience. Yeah, if someone else wants to take over and finish this off, that'd be great!

lemmih commented:
> It would be really nice to have this in. Although it fails to generate a plot for logarithmic scale (min yrange can't be zero). That should can be solved by removing `set(Range::Limits(0, max_formatted[0])` (line 181 from summary.rs). If this is still of interest to @bheisler I could take over this and finish implementing the feature.

I'm ready to review if you take over this PR.

lemmih commented:
Closing due to inactivity.
